### PR TITLE
geo/geomfn: implement ST_MinimumBoundingCircle

### LINKED ISSUE #48987: geo/geomfn: implement ST_MinimumBoundingCircle({geometry,int4})
Implement `ST_MinimumBoundingCircle` on arguments {geometry,int4}, which should adopt [PostGIS behaviour](https://postgis.net/docs/ST_MinimumBoundingCircle.html).

*Observers: Please react to this issue if you need this functionality.*

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo/geomfn`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo/geomfn) (parse and output related functions can go in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo)). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output. Tests here should just ensure the builtin is linked end to end (your exhaustive unit tests go the above mentioned packages!).
* Ensure the documentation is regenerated by calling `make buildshort`. You can also play with it by calling `./cockroach demo --empty` afterwards.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48552 for an example PR.

The following additional guidance has been issued on implementing this function:

Implemented by GEOS.

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2020-08-30T17:39:33Z. Changes to titles, body and labels may be overwritten.</sub>

### PR DESCRIPTION
**Pull request by C0rWin**
This commit implements ST_MinimumBoundingCircle builtin function to
return polygon shape which approximates minimum bounding circle to
contain provided geometry.

Resolves: #48987

Signed-off-by: Artem Barger <bartem@il.ibm.com>

Release note (sql change): Returns polygon shape which approximates
minimum bounding circle to contain geometry

### CODE REVIEW

**Comment by C0rWin** [comment]
Seems to fail logic tests in CI due to precision issues, are there any suggestions how to overcome it, @otan ?

**Review by otan — COMMENTED** [review]
i'll review this one after you rebase with MinimumBoundingRadius when that merges!

**Comment by C0rWin** [comment]
> i'll review this one after you rebase with MinimumBoundingRadius when that merges!

Just rebased on top the previous one, though, still need to wait to merge MinimumBoundingRadius first.

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 1]
```diff
@@ -5539,6 +5539,28 @@ See http://developers.google.com/maps/documentation/utilities/polylinealgorithm`
...
+			Fn: func(ctx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+				geometry := tree.MustBeDGeometry(args[0])
+
+				polygon, _, _, err := geomfn.MinimumBoundingCircle(geometry.Geometry)
+				if err != nil {
+					return nil, err
+				}
+				return tree.NewDGeometry(polygon), nil
+			},
+			Info:       "Returns the smallest circle polygon that can fully contain a geometry. ",
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:5560]: nit: remove trailing space.

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 2]
```diff
@@ -5539,6 +5539,28 @@ See http://developers.google.com/maps/documentation/utilities/polylinealgorithm`
 			Volatility: tree.VolatilityImmutable,
 		}),
 
+	"st_minimumboundingcircle": makeBuiltin(defProps(),
+		tree.Overload{
+			Types: tree.ArgTypes{
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:5544]: nit: use `geometryOverload1` for this

Comment (C0rWin) [reply pkg/sql/sem/builtins/geo_builtins.go]: done

File: pkg/sql/sem/builtins/geo_builtins.go (hunk 1 above)
Comment (C0rWin) [reply pkg/sql/sem/builtins/geo_builtins.go]: done

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 3]
```diff
@@ -5539,6 +5539,19 @@ See http://developers.google.com/maps/documentation/utilities/polylinealgorithm`
 			Volatility: tree.VolatilityImmutable,
 		}),
 
+	"st_minimumboundingcircle": makeBuiltin(defProps(),
+		geometryOverload1(func(evalContext *tree.EvalContext, g *tree.DGeometry) (tree.Datum, error) {
+			polygon, _, _, err := geomfn.MinimumBoundingCircle(g.Geometry)
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:5544]: it looks like we're missing the "num_segs_per_qt_circ" arg here -- see https://postgis.net/docs/ST_MinimumBoundingCircle.html.

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 4]
```diff
@@ -5539,6 +5539,19 @@ See http://developers.google.com/maps/documentation/utilities/polylinealgorithm`
...
+			if err != nil {
+				return nil, err
+			}
+			return tree.NewDGeometry(polygon), nil
+		}, types.Geometry,
+			infoBuilder{
+				info: "Returns the smallest circle polygon that can fully contain a geometry.",
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:5551]: > smallest circle polygon

nit: polygon approximated as a circle

File: pkg/sql/logictest/testdata/logic_test/geospatial [hunk 5]
```diff
@@ -5541,3 +5541,15 @@ SELECT ST_AsText(center), round(radius, 2) FROM ST_MinimumBoundingRadius('POLYGO
...
+----
+POLYGON ((135.597000000000008 115, 134.384999999999991 102.689999999999998, 130.794000000000011 90.853999999999999, 124.963000000000008 79.945000000000007, 117.116 70.384, 107.555000000000007 62.536999999999999, 96.646000000000001 56.706000000000003, 84.810000000000002 53.115000000000002, 72.5 51.902999999999999, 60.189999999999998 53.115000000000002, 48.353999999999999 56.706000000000003, 37.445 62.536999999999999, 27.884 70.384, 20.036999999999999 79.945000000000007, 14.206 90.853999999999999, 10.615 102.689999999999998, 9.403 115, 10.615 127.310000000000002, 14.206 139.146000000000015, 20.036999999999999 150.055000000000007, 27.884 159.616000000000014, 37.445 167.462999999999994, 48.353999999999999 173.294000000000011, 60.189999999999998 176.884999999999991, 72.5 178.097000000000008, 84.810000000000002 176.884999999999991, 96.646000000000001 173.294000000000011, 107.555000000000007 167.462999999999994, 117.116 159.616000000000014, 124.963000000000008 150.055000000000007, 130.794000000000011 139.146000000000015, 134.384999999999991 127.310000000000002, 135.597000000000008 115))
```
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/geospatial:5550]: can you add a test for the minimum bounding circle of a single point?

**Review by otan — COMMENTED** [review]
Hey -- read the docs over and noticed we're missing a permutation of this argument.

File: pkg/sql/sem/builtins/geo_builtins.go (hunk 3 above)
Comment (C0rWin) [reply pkg/sql/sem/builtins/geo_builtins.go:5544]: I have noted this as well, however, didn't find in GEOS implementation where you can control this. Do you have suggestions? 

```c++
extern GEOSGeometry GEOS_DLL *GEOSMinimumBoundingCircle_r(GEOSContextHandle_t handle,
                                                const GEOSGeometry* g, double* radius,
                                                GEOSGeometry** center);
```

File: pkg/sql/logictest/testdata/logic_test/geospatial (hunk 5 above)
Comment (C0rWin) [reply pkg/sql/logictest/testdata/logic_test/geospatial:5550]: yes, sure.

Comment (C0rWin) [reply pkg/sql/logictest/testdata/logic_test/geospatial:5550]: done

File: pkg/sql/sem/builtins/geo_builtins.go (hunk 3 above)
Comment (C0rWin) [reply pkg/sql/sem/builtins/geo_builtins.go:5544]: Looks like I've found how to handle this in [PostGis](https://github.com/postgis/postgis/commit/6a097e28868573af2b2b87084ca809966d5b5dfa#diff-ef8622c09f1c3310d8f5496c7fa4390dd103ea08647ed1ff884ca05bfcdfdf36R6202)

Comment (otan) [reply pkg/sql/sem/builtins/geo_builtins.go:5544]: I was going to suggest using ST_Buffer on the center by the radius.

Comment (C0rWin) [reply pkg/sql/sem/builtins/geo_builtins.go:5544]: > I was going to suggest using ST_Buffer on the center by the radius.

Added implementation to do just this.

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 6]
```diff
@@ -5539,6 +5539,45 @@ See http://developers.google.com/maps/documentation/utilities/polylinealgorithm`
...
 		}),
 
+	"st_minimumboundingcircle": makeBuiltin(defProps(),
+		geometryOverload1(func(evalContext *tree.EvalContext, g *tree.DGeometry) (tree.Datum, error) {
+			polygon, _, _, err := geomfn.MinimumBoundingCircle(g.Geometry)
+			if err != nil {
+				return nil, err
+			}
+			return tree.NewDGeometry(polygon), nil
+		}, types.Geometry,
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:5549]: nit: new line per argument, i.e.

```
geometryOverload1(
  func(evalContext *tree.EvalContext, g *tree.DGeometry) (tree.Datum, error) {
  },
  types,Geometry,
  infoBuilder{
    ...
  },
},
tree.Overload{ , / ...

```

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 7]
```diff
@@ -5539,6 +5539,45 @@ See http://developers.google.com/maps/documentation/utilities/polylinealgorithm`
...
+			Volatility: tree.VolatilityImmutable,
+			Fn: func(evalContext *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+				g := tree.MustBeDGeometry(args[0])
+				numOfSeg := tree.MustBeDInt(args[1])
+				_, centroid, radius, err := geomfn.MinimumBoundingCircle(g.Geometry)
+				if err != nil {
+					return nil, err
+				}
+
+				polygon, err := geomfn.Buffer(centroid,
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:5568]: nit: new line per argument, i.e.

```
polygon, err := geomfn.Buffer(
   centroid,
   geomfn.MakeDefaultBufferParams().WithQuadrantSegments(int(numOfSeg)),
   radius,
)
```

**Review by otan — COMMENTED** [review]
thanks for the update! a small stylistic comment and this is good to go :)

File: pkg/sql/sem/builtins/geo_builtins.go (hunk 6 above)
Comment (C0rWin) [reply pkg/sql/sem/builtins/geo_builtins.go]: done

File: pkg/sql/sem/builtins/geo_builtins.go (hunk 7 above)
Comment (C0rWin) [reply pkg/sql/sem/builtins/geo_builtins.go]: done

**Comment by C0rWin** [comment]
> thanks for the update! a small stylistic comment and this is good to go :)

small update + rebase.

**Review by otan — APPROVED** [review]
thanks!

bors r+

**Comment by otan** [comment]
looks like a lint failure

**Comment by C0rWin** [comment]
> looks like a lint failure

apologies missed that one, fixed now

**Review by otan — APPROVED** [review]
thanks

bors r+

**Pull request merged**